ansible_events_ui/db/utils/lostream.py

```diff
@@ -0,0 +1,227 @@
+        self.pos += len(buff)
+
+        return buff.decode("utf-8") if self.text_data else buff
+
+    async def write(self, buff: Union[str, bytes]) -> int:
+        self.closed_check()
+        if self.imode == self.INV_READ:
+            raise UnsupportedOperation("not writeable")
+
+        if len(buff) > 0:
+            sql = """
+select lo_put(:_oid, :_pos, :_buff) as lo_bytes
+;
+"""
+            if self.text_data and isinstance(buff, str):
+                buff = buff.encode("utf-8")
+
+            bufflen = len(buff)
+            await self.session.execute(
+                sql, {"_oid": self.oid, "_pos": self.pos, "_buff": buff}
+            )
+            self.pos += bufflen
+
+            return bufflen
+        else:
+            return 0
+
+    async def flush(self) -> None:
+        self.closed_check()
+        if self.imode == self.INV_READ:
+            raise UnsupportedOperation("not writeable")
+        await self.session.commit()
+
+    async def truncate(self) -> None:
+        self.closed_check()
+        if self.imode == self.INV_READ:
+            raise UnsupportedOperation("not writeable")
+
+        open_sql = """
+select lo_open(:_oid, :_mode) as lofd;
+        """
+        trunc_sql = """
+select lo_truncate(:_fd, :_len);
+        """
+        close_sql = """
+select lo_close(:_fd) as lofd;
+        """
+        cur = await self.session.execute(
+            open_sql, {"_oid": self.oid, "_mode": self.imode}
+        )
+        fd = cur.first().lofd
+        await self.session.execute(trunc_sql, {"_fd": fd, "_len": self.pos})
+        await self.session.execute(close_sql, {"_fd": fd})
+        self.length = self.pos
+
+    async def close(self) -> None:
+        if (
+            not self.closed
+            and self.oid > 0
+            and (self.imode & self.INV_WRITE)
+            and (self.pos != self.length)
+        ):
+            await self.truncate()
+
+        if self.imode & self.INV_WRITE:
+            await self.flush()
+
+    async def delete(self) -> None:
+        if self.oid is not None and self.oid > 0:
+            await self.session.execute(
+                """
+select lo_unlink(:oid) ;
+                """,
+                {"oid": self.oid},
+            )
+            self.oid = 0
+
+
+async def _create_large_object(session: AsyncSession) -> int:
+    sql = """
+select lo_create(0) as loid;
+    """
+    loid = (await session.execute(sql)).first().loid
+    await session.commit()
+    return loid
+
+
+async def _verify_large_object(oid: int, session: AsyncSession) -> Tuple:
+    sql = """
+select m.oid,
+       sum(length(coalesce(d.data, convert_to('', 'utf-8')))) as _length
+  from pg_largeobject_metadata m
+  left
+  join pg_largeobject d
+    on d.loid = m.oid
+ where m.oid = :_oid
+ group
+   by m.oid;
+    """
+    rec = (await session.execute(sql, {"_oid": oid})).first()
+
+    return (rec.oid, rec._length) if rec else (None, None)
```

Comment:
Please avoid unnecessary use of ternary expressions in Python.
1. They go against Zen of Python:
    ```
    Sparse is better than dense.
    Readability counts.
    ```
2. They don't work well with test coverage tools.

Also as I mentioned before, in Python leading underscore often means an attribute not designed for external access. Sometimes it is enforced by linting tools. There are few exception to that rule, for instance named tuple like object, that have arbitrary user defined set of fields. But in this case and in other cases in the code I don't see why do you use leading underscore prefix for column aliases and placeholders.

Reply:
In this situation, I feel that leaving the if else in a one liner is fine. It's still easy to read and in my experience - the test coverage tools will still require you to cover both cases whether or not we break it out into a full blown if/else statement. Conditional expressions are a part of Python general syntax and it is definitely more readable than C-style ternaries, so exploding it into an if-statement is not really necessary IMO.

Reply:
@abaiken I can accept your opinion on readability, I'll only add that if-expressions are better suited in cases where expression is a requirement (lambda functions for example). Regarding coverage however, it's not that simple. This is an example of how python `coverage` package calculate the code coverage:

https://gist.github.com/cutwater/1bba23b87a96a50304a46b716c568b3f

Reply:
Thanks for the example @cutwater. It helped me understand your point about coverage.